src/plugins/typedispatcher/helpers.c

```diff
@@ -0,0 +1,437 @@
+		}
+	}
+	ksRewind (result);
+	return result;
+}
+
+// return a keyset containing all keys below <key>
+KeySet * getAllKeysBelow (const Key * key, KeySet * ks)
+{
+	return getKeysBelow (key, ks, ELEKTRA_REL_BELOW_SAME_NS);
+}
+
+// return a keyset containing all keys directly below <key>
+KeySet * getKeysDirectBelow (const Key * key, KeySet * ks)
+{
+	return getKeysBelow (key, ks, ELEKTRA_REL_DIRECT_BELOW_SAME_NS);
+}
+
+// check if key is withing scope of datatype
+// key is within scope if either the key is in the same namespace
+// and below the key where the datatype has been defined, or if
+// key is below the key where the datatype has been defined and the
+// datatype has been defined in the spec namespace
+
+int isWithinScope (const TypeConfig * config, const Key * key)
+{
+	const Key * scope = config->scope;
+	if (keyCmp (scope, key))
+	{
+		if (keyRel2 (scope, key, ELEKTRA_REL_BELOW_SAME_NS) <= 0)
+		{
+			elektraNamespace scopeNS = keyGetNamespace (scope);
+			if (scopeNS == KEY_NS_SPEC)
+			{
+				if (keyRel2 (scope, key, ELEKTRA_REL_BELOW_IGNORE_NS) <= 0)
+				{
+					return 0;
+				}
+				else
+				{
+					return 1;
+				}
+			}
+			else
+			{
+				return 0;
+			}
+		}
+		else
+		{
+			return 1;
+		}
+	}
+	else
+	{
+		return 1;
+	}
+}
+
+
+// initialize type config skeleton
+TypeConfig * newTypeConfig ()
+{
+	TypeConfig * tc = NULL;
+	tc = elektraCalloc (sizeof (TypeConfig));
+	if (!tc) return NULL;
+	tc->type = SKEL;
+	tc->scope = keyNew (0, KEY_END);
+	tc->checks = ksNew (0, KS_END);
+	tc->types = ksNew (0, KS_END);
+	tc->params = ksNew (0, KS_END);
+	return tc;
+}
+
+// sets TypeConfig->type to SUMTYPE if keyString matches "SUM"
+// SUBTYPE otherwise
+void setTypeType (TypeConfig * tc, const Key * key)
+{
+	if (!strncasecmp (keyString (key), "SUM", 3))
+		tc->type = SUMTYPE;
+	else
+		tc->type = SUBTYPE;
+}
+
+TypeType getTypeType (TypeConfig * tc)
+{
+	return tc->type;
+}
+
+Key * getTypeKey (DispatchConfig * config, const char * type)
+{
+	Key * lookup = ksLookupByName (config->types, type, KDB_O_NONE);
+	return lookup;
+}
+
+// look up <type> and return its configuration if found
+TypeConfig * getType (DispatchConfig * config, const char * type)
+{
+	Key * lookup = getTypeKey (config, type);
+	if (!lookup) return NULL;
+	TypeConfig * tc = *(TypeConfig **)keyValue (lookup);
+	return tc;
+}
+
+char * replaceParametersWithArguments (const Key * key, KeySet * args)
+{
+	const char * ptr = keyString (key);
+#ifdef DEVBUILD
+	fprintf (stderr, "replacing parameters in %s\n", ptr);
+#endif
+	const char * lastPtr = ptr;
+	char * preparedString = NULL;
+	char * dstPtr = NULL;
+	ssize_t preparedStringLen = 1;
+	while (*ptr)
+	{
+		if (*ptr == '%')
+		{
+			const char * endPtr = ptr + 1;
+			if (lastPtr == ptr) ++lastPtr;
+			while (*endPtr)
+			{
+				if (*endPtr == '%')
+				{
+					// copy string from lastPtr to ptr
+					preparedStringLen += (ptr - lastPtr) + 1;
+					elektraRealloc ((void **)&preparedString, preparedStringLen);
```

Comment:
i think we should always include a failure handling for realloc (if (elektraRealloc(..) < 0) ...)